rasa/core/policies/rule_policy.py

```diff
@@ -519,6 +519,37 @@ def _handling_loop_sources(domain: Domain) -> Set[Text]:
             )
         return loop_sources
 
+    def _should_delete(
+        self,
+        prediction_source: Text,
+        tracker: TrackerWithCachedStates,
+        gold_action_name: Text,
+    ) -> bool:
+        """Checks whether this contradiction is due to action, intent pair.
+
+        Args:
+            prediction_source: the states that result in the prediction
+            tracker: the tracker that raises the contradiction
+
+        Returns:
+            true if the contradiction is a result of an action, intent pair in the rule.
+        """
+        if (
+            tracker.is_rule_tracker  # only apply to contradicting story, not rule
+            or prediction_source.count(PREVIOUS_ACTION)
+            > 1  # only apply for prediction after unpredictable action
+            or gold_action_name
+            != ACTION_LISTEN_NAME  # only apply for prediction of action_listen
+        ):
+            return False
+        for source in self.lookup[RULES]:
+            if (
+                source.startswith(prediction_source[:-2])
+                and not prediction_source == source
```

Comment:
could you please also add comments, explaining this

Reply:
Yep! I also added a test to test this, one to test `prediction_source.count(PREVIOUS_ACTION) > 1`, and one to test only applying this for stories. I think there is no point in testing not deleting for not action listen since if this broke, contradicting rules would break?

Reply:
Not sure if I should combine the tests with mark parametrize or if that just makes them less readable because each have a different setup

Reply:
I think separate tests are better

Reply:
> I think there is no point in testing not deleting for not action listen since if this broke, contradicting rules would break?

what do you mean? I think given how the code is organized, it'd be good to test that second non listen action raises contradiction

Reply:
Not sure what you mean? I was saying that I wrote tests for all the conditions in the if blocks of `_should_delete` (is a story, isn't after a predictable action, isn't the last action of the rule...) except for "is action_listen":
```
rule:
- action: A
- action: B #(NOT action_listen)
```
```
story:
- action: A
- action: C
```
(in this case, the `should_delete` method returns false because the contradicting action is not `action_listen`)
Reason I say this it's not necessary to test is because this is already covered by the test case `test_contradicting_rules`.

Reply:
But now that I think about it, myabe it's good to have it anyways for symmetry. Was this what you meant by
> it'd be good to test that second non listen action raises contradiction
?

Reply:
yes, but if it is a duplicate, then no need to add it

Reply:
I think it's not quite a duplicate, because `test_contradicting_rules_and_stories` has:
```
rule:
- intent
- action
- action_listen
```

and the test I added specifically has no action_listen